### ISSUE TITLE
When does node-pg use binary protocol

### ISSUE THREAD

**Issue by isaacl** [brianc/node-postgres#2500]
Hi! I was looking at [node-pg-types](brianc/node-pg-types) and noticed it has a separate file for decoding binary streams  

But when I set a custom type parser, such as on int8 or timestamptz, this only affects the text protocol.  Are there situations where node-pg switches to binary protocol? And in binary protocol, are custom type parsers ignored?

**Comment by charmander** [comment]
> Are there situations where node-pg switches to binary protocol?

I think queries accept a `binary: true` option? It also used to when a `Buffer` parameter was passed; not sure if it still does.

> And in binary protocol, are custom type parsers ignored?

Custom text type parsers don’t apply in that case, right. You can add custom binary parsers with [the three-argument form of `setTypeParser`](https://github.com/brianc/node-pg-types/blob/8594bc6befca3523e265022f303f1376f679b5dc/index.js#L31-L34).

**Comment by isaacl** [comment]
Thanks. My read of node-pg connection is that it does not use the binary protocol, nor is that option documented anymore!

I was looking at [asyncpg](https://github.com/MagicStack/asyncpg) and wondering why we didn't have something similar for node...

**Comment by isaacl** [comment]
![sad perf](https://raw.githubusercontent.com/MagicStack/asyncpg/master/performance.png)

**Comment by brianc** [comment]
Oh interesting I haven't looked at the binary protocol in a while, I'll
take a look at it and maybe there are some nice perf gains we can get there!

**Comment by JaneJeon** [comment]
As far as I can tell binary mode is not supported yet? https://github.com/brianc/node-postgres/blob/6121bd3bb0e0e8ef8ec8ad5d02f59fef86b2f992/packages/pg-protocol/src/parser.ts#L87

Would be extremely interested in keeping up with this development.

**Comment by isaacl** [comment]
To achieve performance similar in the graph probably just requires porting asyncpg to node.  They rewrote everything and use native code.

**Comment by JaneJeon** [comment]
“just” rewriting everything and writing native code probably is WAY out of scope for this project, not to mention insanely time-consuming for the author. Trying to optimize parts, enabling binary protocol where possible, etc, would more likely result in smaller but easier “wins” that are much more realistic to achieve without a full rewrite.

**Comment by isaacl** [comment]
I get that.  I wasn't saying it was in the scope of this bug. I was saying
we can't expect to reach those performance targets in this project.

Isaac

**Comment by punkpeye** [comment]
Was there ever any progress here?

I am wondering specifically around handling of jsonb.

Parsing of large JSON documents is the current bottleneck I am dealing with. It sounds like using binary protocol would help here.

**Comment by punkpeye** [comment]
Would using pg-native leverage the binary protocol?

**Comment by punkpeye** [comment]
Interestingly, this project claims to use binary protocol:

https://github.com/panates/postgrejs

> Binary Wire Protocol: Implements the full binary wire protocol for all PostgreSQL data types, ensuring robust and efficient data handling.

Never heard of it.

**Comment by charmander** [comment]
> Would using pg-native leverage the binary protocol?

No, pg needs to parse and serialize between JavaScript values and binary representations regardless of whether it’s using libpq.

> Parsing of large JSON documents is the current bottleneck I am dealing with. It sounds like using binary protocol would help here.

https://github.com/brianc/node-postgres/issues/3467#issuecomment-3530402820 would help determine if that’s the case.
